src/Data/IxSet/Typed.hs

```diff
@@ -514,22 +519,47 @@ insert = change Set.insert Ix.insert
 delete :: Indexable ixs a => a -> IxSet ixs a -> IxSet ixs a
 delete = change Set.delete Ix.delete
 
+-- | Internal implementation for update* family
+updateIx' :: (Indexable ixs a, IsIndexOf ix ixs)
+         => (IxSet ixs a -> Maybe a) -> ix -> a -> IxSet ixs a -> IxSet ixs a
+updateIx' get i new ixset = insert new $
+                     maybe ixset (flip delete ixset) $
+                     get $ ixset @= i
+
+-- | Internal implementation for delete* family
+deleteIx' :: (Indexable ixs a, IsIndexOf ix ixs)
+         => (IxSet ixs a -> Maybe a) -> ix -> IxSet ixs a -> IxSet ixs a
+deleteIx' get i ixset = maybe ixset (flip delete ixset) $
+                       get $ ixset @= i
+
 -- | Will replace the item with the given index of type 'ix'.
 -- Only works if there is at most one item with that index in the 'IxSet'.
 -- Will not change 'IxSet' if you have more than one item with given index.
 updateIx :: (Indexable ixs a, IsIndexOf ix ixs)
          => ix -> a -> IxSet ixs a -> IxSet ixs a
-updateIx i new ixset = insert new $
-                     maybe ixset (flip delete ixset) $
-                     getOne $ ixset @= i
+updateIx = updateIx' getOne
+
+-- | Will replace the item with the given index of type 'ix'.
+-- Only works if there is at most one item with that index in the 'IxSet'.
+-- Will generate an error if there is more than one item with given index.
+updateUnique :: (Indexable ixs a, IsIndexOf ix ixs)
+         => ix -> a -> IxSet ixs a -> IxSet ixs a
```

Comment:
I feel like API wise this is a bit rough. The caller may want to decide whether he wants to error on non-unique index. Options are:
* wrap it in `Maybe (IxSet ixs s)` or `Either (IxSet ixs s) (IxSet ixs s)`
* return a tuple `(Bool, IxSet ixs a)`
* create a dedicated type with a record selector for people who don't care about whether the update worked or not

Reply:
Ok, its been updated to use MonadThrow